lib/web_ui/lib/src/engine/canvaskit/render_canvas.dart

```diff
@@ -0,0 +1,165 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:ui/ui.dart' as ui;
+
+import '../dom.dart';
+import '../window.dart';
+import 'renderer.dart';
+import 'util.dart';
+
+/// A visible (on-screen) canvas that can display bitmaps produced by CanvasKit
+/// in the (off-screen) SkSurface which is backed by an OffscreenCanvas.
+///
+/// In a typical frame, the content will be rendered via CanvasKit in an
+/// OffscreenCanvas, and then the contents will be transferred to the
+/// RenderCanvas via `transferFromImageBitmap()`.
+///
+/// If we need more RenderCanvases, for example in the case where there are
+/// platform views and we need overlays to render the frame correctly, then
+/// we will create multiple RenderCanvases, but crucially still only have
+/// one OffscreenCanvas which transfers bitmaps to all of the RenderCanvases.
+///
+/// To render into the OffscreenCanvas with CanvasKit we need to create a
+/// WebGL context, which is not only expensive, but the browser has a limit
+/// on the maximum amount of WebGL contexts which can be live at once. Using
+/// a single OffscreenCanvas and multiple RenderCanvases allows us to only
+/// create a single WebGL context.
+class RenderCanvas {
+  RenderCanvas();
+
+  /// The root HTML element for this canvas.
+  ///
+  /// This element contains the canvas used to draw the UI. Unlike the canvas,
+  /// this element is permanent. It is never replaced or deleted, until this
+  /// canvas is disposed of via [dispose].
+  ///
+  /// Conversely, the canvas that lives inside this element can be swapped, for
+  /// example, when the screen size changes, or when the WebGL context is lost
+  /// due to the browser tab becoming dormant.
+  final DomElement htmlElement = createDomElement('flt-canvas-container');
+
+  /// The underlying `<canvas>` element used to display the pixels.
+  DomCanvasElement? canvasElement;
+  int _pixelWidth = -1;
+  int _pixelHeight = -1;
+
+  DomCanvasRenderingContextBitmapRenderer? renderContext;
+
+  ui.Size? _currentCanvasPhysicalSize;
+  ui.Size? _currentRenderSize;
+  double _currentDevicePixelRatio = -1;
+
+  bool _addedToScene = false;
+  void addToScene() {
```

Comment:
It looks suspicious to me that the canvas knows where to add itself among other canvases and platform views in the scene host. It's like `int` knowing how to add itself to a `List<int>`.

Suggestion: can we move this logic to the place that makes decisions about how render canvases and platform views are composited?

Reply:
Done.